kvserver: clean up replica unquiescence

erikgrinaker opened the pull request:
**kvserver: remove stale `maybeQuiesce` TODO**

The leader won't quiesce if followers aren't caught up.

**kvserver: remove `maybeUnquiesceAndWakeLeaderLocked`**

This patch merges `maybeUnquiesceAndWakeLeaderLocked()` into `maybeUnquiesceWithOptionsLocked()`, using a separate parameter to wake the leader. Care is taken to make this purely mechanical, with no logical changes at all.

**kvserver: remove `maybeUnquiesceWithOptionsLocked`**

This patch merges `maybeUnquiesceWithOptionsLocked()` into `maybeUnquiesceLocked()`, requiring callers to always specify options. The subtlety around unquiescence, and few call sites, makes it beneficial to be explicit.

This is a purely mechanical change, with no logical changes. A couple of tests have been changed to now wake the leader when unquiescing, but this has no bearing on the tests.

**kvserver: set last update times on leader unquiesce**

Previously, `Replica.lastUpdateTimes` was updated whenever a replica unquiesced without attempting to wake the leader. However, this had two flaws: a leader could fail to call it if it hit a code path where it did attempt to wake the leader (even if it was leader itself), e.g. by returning `true` from `withRaftGroup()`, and it could also be called on a follower where it would have no effect.

This patch instead updates it when unquiescing the leader, regardless of unquiesce options.

**kvserver: only wake leader when unquiescing a follower**

Previously, any replica would wake the leader when unquiescing, if requested by the caller. However, this could cause the leader to propose an empty command to wake itself, which commonly happens in `handleRaftReady()` via `withRaftGroupLocked()`. This appears unnecessary, and likely causes a large amount of Raft proposals with ranges that frequently (un)quiesce.

This patch instead only attempts to wake the leader from followers.

**kvserver: wake leader before campaigning when unquiescing**

We should wake the leader before campaigning when unquiescing, since we won't send the proposal in the candidate state and thus won't give the leader a chance to assert leadership if we're wrong about it being dead.

**kvserver: omit unnecessary campaign checks on leader unquiescence**

This patch changes a couple of `maybeUnquiesce()` call sites to not attempt to campaign when we know the replica must already be the leader.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: remove `maybeUnquiesceAndWakeLeaderLocked`

This patch merges `maybeUnquiesceAndWakeLeaderLocked()` into
`maybeUnquiesceWithOptionsLocked()`, using a separate parameter to wake
the leader. Care is taken to make this purely mechanical, with no
logical changes at all.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: remove `maybeUnquiesceWithOptionsLocked`

This patch merges `maybeUnquiesceWithOptionsLocked()` into
`maybeUnquiesceLocked()`, requiring callers to always specify options.
The subtlety around unquiescence, and few call sites, makes it
beneficial to be explicit.

This is a purely mechanical change, with no logical changes. A couple of
tests have been changed to now wake the leader when unquiescing, but
this has no bearing on the tests.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: set last update times on leader unquiesce

Previously, `Replica.lastUpdateTimes` was updated whenever a replica
unquiesced without attempting to wake the leader. However, this had two
flaws: a leader could fail to call it if it hit a code path where it did
attempt to wake the leader (even if it was leader itself), e.g. by
returning `true` from `withRaftGroup()`, and it could also be called on
a follower where it would have no effect.

This patch instead updates it when unquiescing the leader, regardless of
unquiesce options.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: remove stale `maybeQuiesce` TODO

The leader won't quiesce if followers aren't caught up.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: only wake leader when unquiescing a follower

Previously, any replica would wake the leader when unquiescing, if
requested by the caller. However, this could cause the leader to propose
an empty command to wake itself, which commonly happens in
`handleRaftReady()` via `withRaftGroupLocked()`. This appears
unnecessary, and likely causes a large amount of Raft proposals with
ranges that frequently (un)quiesce.

This patch instead only attempts to wake the leader from followers.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: omit unnecessary campaign checks on leader unquiescence

This patch changes a couple of `maybeUnquiesce()` call sites to not
attempt to campaign when we know the replica must already be the leader.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105041)
<!-- Reviewable:end -->

erikgrinaker commented:
@tbg @pavelkalinnikov Would like some scrutiny of the "only wake leader when unquiescing a follower" commit here. It seems problematic and unnecessary that we submit a proposal when the leader wakes itself, but it's possible that we've come to rely implicitly on this logic. I expect some test fallout from this as well.

erikgrinaker commented:
This resulted in a substantial improvement when running kv0 on 20k ranges with `COCKROACH_QUIESCE_AFTER_TICKS=0` where we aggressively quiesce ranges (the default in 23.1). The Raft proposal rate was cut almost by half, and the Raft message rate dropped by 30%:

<img width="1055" alt="Screenshot 2023-06-16 at 16 50 59" src="https://github.com/cockroachdb/cockroach/assets/644420/d44a5fff-0624-49be-a078-340a4e1ce306">

Throughtput improved by 6% (sample size of 1 though):

```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  180.0s        0        1840221        10223.4     18.8     15.2     50.3     71.3    520.1  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  180.0s        0        1953950        10855.3     17.7     13.6     52.4     71.3    469.8  write
```

I'm not sure that this is safe yet, since I saw a fair bit of election timeouts after cold starting the cluster. Having a closer look.

erikgrinaker commented:
>I'm not sure that this is safe yet, since I saw a fair bit of election timeouts after cold starting the cluster. Having a closer look.

Saw the same problem with `master`, so it isn't this PR, but we may have broken something recently. Digging in.

erikgrinaker commented:
> we may have broken something recently.

#105052

erikgrinaker committed:
kvserver: wake leader before campaigning when unquiescing

We should wake the leader before campaigning when unquiescing, since we
won't send the proposal in the candidate state and thus won't give the
leader a chance to assert leadership if we're wrong about it being dead.

Epic: none
Release note: None

tbg reviewed (approved):
Nice, thanks.

pavelkalinnikov commented on pkg/kv/kvserver/replica_raft_quiesce.go:
```diff
@@ -75,34 +86,28 @@ func (r *Replica) maybeUnquiesceWithOptionsLocked(campaignOnWake bool) bool {
+	if wakeLeader {
+		// Propose an empty command which will wake the leader.
+		if log.V(3) {
+			log.Infof(ctx, "waking %d leader", r.RangeID)
```
How about "r%d" instead of "%d". It's easier to search by "rNNN" in logs when investigating, this avoids false positives.

erikgrinaker replied:
Tacked on a commit.

pavelkalinnikov commented on pkg/kv/kvserver/helpers_test.go:
```diff
@@ -495,15 +495,7 @@ func (r *Replica) GetQueueLastProcessed(ctx context.Context, queue string) (hlc.
 }
 
 func (r *Replica) MaybeUnquiesce() bool {
```
The semantics of `MaybeUnquiesce` changed. Were there any callers who wanted the old behaviour (wakeLeader=mayCampaign=false)?

erikgrinaker commented on pkg/kv/kvserver/replica_raft_quiesce.go:
```diff
@@ -94,9 +94,8 @@ func (r *Replica) maybeUnquiesceLocked(wakeLeader, mayCampaign bool) bool {
-	}
 
-	if wakeLeader {
+	} else if st.RaftState == raft.StateFollower && wakeLeader {
```
Correct, a candidate has no leader.

erikgrinaker commented on pkg/kv/kvserver/store_raft.go:
```diff
@@ -705,7 +705,7 @@ func (s *Store) nodeIsLiveCallback(l livenesspb.Liveness) {
 		r.mu.RUnlock()
 		if quiescent && lagging.MemberStale(l) {
-			r.maybeUnquiesce(false /* wakeLeader */, true /* mayCampaign */)
+			r.maybeUnquiesce(false /* wakeLeader */, false /* mayCampaign */) // already leader
```
Because `laggingFollowersOnQuiesce` is only non-nil on the leader.

erikgrinaker committed:
kvserver: use r-prefix for range IDs in quiescence logs

Epic: none
Release note: None

erikgrinaker commented:
TFTRs!

bors r+
